### PR TITLE
Create a Quartz Extension Scheduler with Typed Akka Actor.

### LINKED ISSUE #101: Support for Akka Typed
As the typed akka api is now the default in the akka project it would be nice if akka-quartz-scheduler would support it.

For example all api's that currently have the current format:

``` scala
def schedule(name: String, receiver: ActorRef, msg: AnyRef)
```

should have an additional typed implemention of the form

``` scala
def schedule[T](name: String, receiver: ActorRef[T], msg: T)
```

 However it would require a typed implementation of SimpleActorMessageJob I think. My first thought for the content of the jobDataMap would be:

```scala
case class ReceiverWithMessage[T](receiver: ActorRef[T], msg: T)
val jobDataMap = Map[String, AnyRef](
      "logBus" -> system.eventStream,
      "receiverWithMessage" -> ReceiverWithMessage(receiver, msg)
    )
```

However I am currently not sure what to do with MessageRequireFireTime.

### PR DESCRIPTION
**Pull request by felipebonezi**
Closes #101

Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

### CODE REVIEW

**Comment by felipebonezi** [comment]
@enragedginger how do you think about this code?
I need to use Quartz with Typed Actors in my project.

**Comment by froth** [comment]
Thanks for your effort!

This basically works by converting ActorRef[A] to ActorRef[Any] to the old nontyped ActorRef, right? This works because of the compatibility layer of akka to support mixing TypedActors and "old" Actors. I think it would work for now (until someone deprecates the compatibility). 

It would be nice if we could have something that supports typed actors all the way through the implementation. But I think this is a great first step. And maybe it really is enough and I am overthinking it ;)

**Comment by felipebonezi** [comment]
> Thanks for your effort!
> 
> This basically works by converting ActorRef[A] to ActorRef[Any] to the old nontyped ActorRef, right? This works because of the compatibility layer of akka to support mixing TypedActors and "old" Actors. I think it would work for now (until someone deprecates the compatibility). 
> 
> It would be nice if we could have something that supports typed actors all the way through the implementation. But I think this is a great first step. And maybe it really is enough and I am overthinking it ;)

I understand your position and agree with it, but we can deliver faster to anyone who want to use until we work on a new major version.

I'll write unit test code and open the PR waiting for approval.

File: build.sbt
Comment (enragedginger) [inline build.sbt:12]: Can you set the versions on the current akka dependencies to `2.6.10` as well?

File: build.sbt
Comment (enragedginger) [inline build.sbt:13]: Can you mark this as `provided`?

File: src/main/scala/QuartzSchedulerTypedExtension.scala
Comment (enragedginger) [inline src/main/scala/QuartzSchedulerTypedExtension.scala:7]: I don't think you get much by extending `ExtensionId[QuartzSchedulerExtension]`. Wouldn't this be better as `ExtensionId[QuartzSchedulerTypedExtension]`? You'd need to override `lookup` and `get` too.

File: src/main/scala/QuartzSchedulerTypedExtension.scala
Comment (enragedginger) [inline src/main/scala/QuartzSchedulerTypedExtension.scala:31]: `msg` should be of type `T` on these, correct?

**Review by enragedginger — CHANGES_REQUESTED** [review]
Hello @felipebonezi,

Thanks for putting this together. I have a handful of comments.

Admittedly, I'm largely unfamiliar with Akka typed, so thanks for taking up this work. Can you add some tests for the Akka typed API? Also, can you update the README to include a section about the typed API for this project and how to use it?

File: build.sbt
Comment (felipebonezi) [reply build.sbt:12]: Sure, I didn't set it to `2.6.10` because I was afraid to broke something. 😆

File: src/main/scala/QuartzSchedulerTypedExtension.scala
Comment (felipebonezi) [reply src/main/scala/QuartzSchedulerTypedExtension.scala:7]: @enragedginger I've implemented the `get` method, but `lookup` doesn't exist in `Akka Typed`.
Can you check this info? Even the documentation is implementing without `lookup` method.

Captura de Tela 2020-11-18 às 12 39 36

**Comment by felipebonezi** [comment]
@enragedginger Why `travis.yml` hasn't a `script` section to compile and test this project? 🤔 
https://github.com/felipebonezi/akka-quartz-scheduler/blob/feature/akka-actor-typed/.travis.yml

**Comment by felipebonezi** [comment]
@enragedginger anything to improve?

File: src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala
Comment (enragedginger) [inline src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala:366]: This object appears to be completely duplicated between the two test packages, no? Can you move it out to its own file and update both test classes to point to it?

File: src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala
Comment (enragedginger) [inline src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala:70]: All of the tests in this class appear to be duplicates of what's in the other test class for the other version of the API. We definitely want to test all of the same cases for both APIs.

My hesitation on approving this PR sooner was that I wanted to look and see if there was a way to avoid duplicating this code. Do you think it's possible to refactor these tests out to either (A) a base testing `trait` or (B) some kind of stateless testing class? Do you see a good way of doing this?

**Review by enragedginger — CHANGES_REQUESTED** [review]

File: src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala
Comment (felipebonezi) [reply src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala:70]: I'll think about it how to do it...

Comment (felipebonezi) [reply src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala:70]: @enragedginger hey, good afternoon!
I analyzed the point you commented, but note that only the logic is repeated, but the classes are not. They use classes that have no relationship with each other (there is no generation or polymorphism between them). Therefore, I have no way of creating a Trait that can handle classes of different types because unused methods are used.

I recommend that you continue as you are, because you ensure that the same tests as the old version of Akka Actor works with the Akka Typed Actor. In this way, we will show the community that, just as it worked in the past, will work in the future.

What do you think?

File: src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala
Comment (felipebonezi) [reply src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala]: Fixed ✅

File: src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala
Comment (felipebonezi) [reply src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala:67]: @enragedginger is there any news about this PR?
Did you consider what I've said?

**Review by enragedginger — APPROVED** [review]
LGTM. Thanks for making this contribution! My apologies on the slow review.

File: src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala
Comment (enragedginger) [reply src/test/scala/QuartzTypedSchedulerFunctionalSpec.scala:67]: Yeah, that makes sense. Thanks for looking into it!

**Pull request merged**
